pkg/log/logger.go

```diff
@@ -0,0 +1,295 @@
+	if cfg, active, emit := l.config(level); emit {
+		active.Log(LevelInfo, cfg.source(), format, args...)
+	}
+}
+
+// Warn logs a warning message.
+func (l logger) Warn(format string, args ...interface{}) {
+	level := LevelWarn
+	if cfg, active, emit := l.config(level); emit {
+		active.Log(level, cfg.source(), format, args...)
+	}
+}
+
+// Error logs an error message.
+func (l logger) Error(format string, args ...interface{}) {
+	level := LevelError
+	if cfg, active, emit := l.config(level); emit {
+		active.Log(level, cfg.source(), format, args...)
+	}
+}
+
+// Fatal logs a fatal error message and os.Exit(1)'s.
+func (l logger) Fatal(format string, args ...interface{}) {
+	level := LevelFatal
+	cfg, active, _ := l.config(level)
+	active.Log(level, cfg.source(), format, args...)
+
+	os.Exit(1)
+}
+
+// Panic logs a panic message and panic()'s.
+func (l logger) Panic(format string, args ...interface{}) {
+	level := LevelPanic
+	cfg, active, _ := l.config(level)
+	active.Log(level, cfg.source(), format, args...)
+
+	panic(fmt.Sprintf(cfg.source()+" "+format, args...))
+}
+
+// DebugBlock logs a multi-line debug message.
+func (l logger) DebugBlock(prefix string, format string, args ...interface{}) {
+	level := LevelDebug
+	if cfg, active, emit := l.config(level); emit {
+		active.Block(level, cfg.source(), prefix, format, args...)
+	}
+}
+
+// InfoBlock logs a multi-line informational message.
+func (l logger) InfoBlock(prefix string, format string, args ...interface{}) {
+	level := LevelInfo
+	if cfg, active, emit := l.config(level); emit {
+		active.Block(level, cfg.source(), prefix, format, args...)
+	}
+}
+
+// WarnBlock logs a multi-line warning message.
+func (l logger) WarnBlock(prefix string, format string, args ...interface{}) {
+	level := LevelWarn
+	if cfg, active, emit := l.config(level); emit {
+		active.Block(level, cfg.source(), prefix, format, args...)
+	}
+}
+
+// ErrorBlock logs a multi-line error message.
+func (l logger) ErrorBlock(prefix string, format string, args ...interface{}) {
+	level := LevelError
+	if cfg, active, emit := l.config(level); emit {
+		active.Block(level, cfg.source(), prefix, format, args...)
+	}
+}
+
+// config returns the logger's configuration and if the level is logged.
+func (l logger) config(level Level) (config, Backend, bool) {
+	if level != LevelDebug && level < log.level {
+		return config{}, nil, false
+	}
+
+	log.RLock()
+	cfg, _ := log.configs[l]
+	active := log.active
+	forced := log.forced
+	log.RUnlock()
+
+	switch level {
+	case LevelInfo:
+		return cfg, active, cfg.isLogging()
+	case LevelDebug:
+		return cfg, active, cfg.isTracing() || forced
+	default:
+		return cfg, active, true
+	}
+}
+
+//
+// Runtime configuration of a single logger instance.
+//
+
+const (
+	maxLoggers = math.MaxUint16
+	loggingBit = (1 << iota)
+	tracingBit
+)
+
+// config is the configuration parameters for a single logger stuffed into a single uint64
+type config struct {
+	id     uint16
+	enable uint16
+}
```

Comment:
Hmm, at a quick glance this bitwise logging config handling somehow feels a bit too complex. Would it be possible to get rid of this with function pointers (set to nil if the corresponding log level/functionality is disabled) or someting, WDYT?

Reply:
I'd rather not. IMO, it's rather subjective/questionable whether `if foo.fn != nil` is any clearer/simpler than `if foo.isDebugging()`.

Reply:
That was just a quick thought/gut feeling. And yes, the nil pointer was probably a bad idea. An alternative would be using function pointers and setting the pointer to point to a stub when disabling logging. That could possibly eliminate all if's and hopefully the config type.

But, no need to get stuck with this here. I think I can live with this 😄 We could create a new issue (assigned to me?) about this and try to address this separately.